src/VisualStudio/Core/Impl/CodeModel/IProjectCodeModelProvider.cs

```diff
@@ -4,6 +4,6 @@ namespace Microsoft.VisualStudio.LanguageServices.Implementation.CodeModel
 {
     internal interface IProjectCodeModelProvider
     {
-        ProjectCodeModel ProjectCodeModel { get; }
+        IProjectCodeModel ProjectCodeModel { get; }
```

Comment:
is there a benefit to having this interface instead of just accepting that there is the single ProjectCodeModel type?

Reply:
ProjectCodeModel lives in our ServicesVisualStudioImpl project along with the rest of the CodeModel implementation. This is because of some .NET COM marshalling behaviors, which requires the types to be public. We don't want those to be public API, so we hide them in this assembly.

IProjectCodeModel and the consumers thereof live in ServicesVisualStudio which is our public type. That's where AbstractProject is, and that can't be moved without breaking TypeScript/F#. You see the casting because the implementation is presuming that it's the only practical implementation of it. The right thing here is to have an IProjectCodeModelFactory that maintains a project ID -> ProjectCodeModel type (thereby getting rid of the field on AbstractProject) which would also get rid of casting. I'm not biting that here.

And some context: this is a cleanup change that's leading up to a larger refactoring of the project system bits to expose a core "I want a project in VisualStudioWorkspace" that is a better API that can be used by CPS, but also by consumers like TypeScript or F#. It'll obsolete the need for inheriting from AbstractProject or other things that keep biting us. My hope is this also becomes a public API for "I want a virtual C# project for <insert reason here>" -- for example, this could get rid of the legacy hacks Venus uses today to create projects that support projection buffer scenarios. If somebody wants a new project, and they want to hook up code model to it, I can imagine IProjectCodeModel (with a better name) becoming a public type too, which they'd need to create from some factory for hookup.